lib/checks.sh

```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/env bash
+
+function checks::ensure_supported_stack() {
+	local stack="${1}"
+
+	case "${stack}" in
+		heroku-20 | heroku-22 | heroku-24)
+			return 0
+			;;
+		cedar* | heroku-16 | heroku-18)
+			# This error will only ever be seen on non-Heroku environments, since the
+			# Heroku build system rejects builds using EOL stacks.
+			output::error <<-EOF
+				Error: The '${stack}' stack is no longer supported.
+
+				This buildpack no longer supports the '${stack}' stack since it has
+				reached its end-of-life:
+				https://devcenter.heroku.com/articles/stack#stack-support-details-for-apps-using-classic-buildpacks
+
+				Upgrade to a newer stack to continue using this buildpack.
+			EOF
+			meta_set "failure_reason" "stack::eol"
+			exit 1
+			;;
+		*)
+			output::error <<-EOF
+				Error: The '${stack}' stack is not recognised.
+
+				This buildpack does not recognise or support the '${stack}' stack.
+
+				If '${stack}' is a valid stack, make sure that you are using the latest
+				version of this buildpack and have not pinned to an older release:
+				https://devcenter.heroku.com/articles/managing-buildpacks#view-your-buildpacks
+				https://devcenter.heroku.com/articles/managing-buildpacks#classic-buildpacks-references
+			EOF
+			meta_set "failure_reason" "stack::unknown"
+			exit 1
+			;;
+	esac
+}
+
+# TODO: Turn this into an error in January 2025.
+function checks::warn_if_duplicate_python_buildpack() {
+	local build_dir="${1}"
+
+	# The check for the `PYTHONHOME` env var prevents this warning triggering in the case
+	# where the Python install was committed to the Git repo (which will be handled later).
+	# (The env var can only have come from the `export` file of an earlier buildpack,
+	# since app provided config vars haven't been exported to the environment here.)
+	if [[ -f "${build_dir}/.heroku/python/bin/python" && -v PYTHONHOME ]]; then
+		output::warning <<-EOF
+			Warning: The Python buildpack has already been run this build.
+
+			An existing Python installation was found in the build directory
+			from a buildpack run earlier in the build.
+
+			This normally means there are duplicate Python buildpacks set
+			on your on your app, which is not supported since it can cause
+			unexpected errors and slow down builds.
+
+			Check the buildpacks set on your app and remove any duplicate
+			Python buildpack entries:
+			https://devcenter.heroku.com/articles/managing-buildpacks#view-your-buildpacks
+			https://devcenter.heroku.com/articles/managing-buildpacks#remove-classic-buildpacks
+
+			If you have a use-case that requires duplicate buildpacks,
+			please comment on:
+			https://github.com/heroku/heroku-buildpack-python/issues/1704
+
+			In January 2025 this warning will be made an error.
```

Comment:
Not blocking comment: This is a pretty aggressive time scale, especially for devs out on holidays and not actively deploying. I feel like we are validated in encouraging people to not do this though, so I'm for an accelerated timescale.

When we change the error to a warning, I would like an env var as an "escape valve" to accomadate not blocking people who (may) have a valid reason we've not yet thought about but haven't had a chance to comment on the issue. For example: An env var to turn the error into a warning or directing people to use an older version of the biuldpack (though that risks people being accidentally encouraged to get stuck on an old buildpack version if they forget about it). A thought.

Reply:
I'll be basing the eventual decision on the additional metics.

However, I already know this is causing real errors (see #1704), and also know very few apps run duplicate buildpacks ([dataclip](https://data.heroku.com/dataclips/rxotsanetozvjufoqvdeyzcpsdcz)).